ff-rewrite-typescript-id-allocator/src/utilities.ts

```diff
@@ -139,3 +141,58 @@ export function uuidStringFromBytes(uuidBytes: Uint8Array | undefined): string |
 	}
 	return uuidString;
 }
+
+// xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx
+const versionMask = 0x4n << (19n * 4n); // Version 4
+const variantMask = 0x8n << (15n * 4n); // Variant RFC4122 (1 0 x x)
+const upperMask = 0xffffffffffffn << (20n * 4n);
+// Upper mask when version/variant bits are removed
+const strippedUpperMask = upperMask >> 6n;
+const middieBittiesMask = 0xfffn << (16n * 4n);
+// Middie mask when version/variant bits are removed
+const strippedMiddieBittiesMask = middieBittiesMask >> 2n;
+// Note: leading character should be 3 to mask at 0011
+// The more-significant half of the N nibble is used to denote the variant (10xx)
+const lowerMask = 0x3fffffffffffffffn;
+// Used to help with stringifying bigints which would otherwise drop trailing zeros
+const precisionMask = 0x1n << 128n;
+
+export function numericUuidFromStableId(stableId: StableId): NumericUuid {
+	const uuidU128 = BigInt(`0x${stableId.replace(/-/g, "")}`);
+	const upperMasked = uuidU128 & upperMask;
+	const middieBittiesMasked = uuidU128 & middieBittiesMask;
+	const lowerMasked = uuidU128 & lowerMask;
+
+	const upperMaskedPlaced = upperMasked >> 6n;
+	const middieBittiesMaskedPlaced = middieBittiesMasked >> 2n;
+
+	const id = upperMaskedPlaced | middieBittiesMaskedPlaced | lowerMasked;
+	return id as NumericUuid;
+}
+
+export function stableIdFromNumericUuid(numericUuid: NumericUuid): StableId {
+	// bitwise reverse transform
+	const upperMasked = (numericUuid & strippedUpperMask) << 6n;
+	const middieBittiesMasked = (numericUuid & strippedMiddieBittiesMask) << 2n;
+	const lowerMasked = numericUuid & lowerMask;
+	const uuidU128 =
+		precisionMask | upperMasked | versionMask | middieBittiesMasked | variantMask | lowerMasked;
+	const uuidString = uuidU128.toString(16).substring(1);
+	return `${uuidString.substring(0, 8)}-${uuidString.substring(8, 12)}-${uuidString.substring(
```

Comment:
Ooh, discuss as well please

Reply:
which line? 181 adds back in the dashes, 180 makes the bigint into a hex string but skips the leading one I had to add to avoid losing trailing 0s (for instance, FFF111 is 000000000000000100010001 in binary, but the bigint would store 100010001 and drop trailing 0s so I padded it with a leading 12**9**th 1 and drop it after to-string).

Reply:
(I called that extra 1 `precisionMask`)